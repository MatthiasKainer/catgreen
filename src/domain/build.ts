export type BuildResult = "unknown" | "failed" | "success" | "running";

export type Build = {
  id: string;
  name: string;
  result: BuildResult;
  meta: {
    triggerReason: string;
    blame: string;
    link: string;
    when: string;
    timestamp: string;
  };
};
