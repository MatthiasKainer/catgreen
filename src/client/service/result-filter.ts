import { Build } from "../../domain/build";

export const objectivedFilter = (...filter: string[]) =>
  filter.reduce(
    (result, current) => {
      const [key, value] = current.split(":");
      if (!value) return result;
      result[key.trim()] = [...(result[key] ?? []), value.trim()];
      return result;
    },
    {} as { [key: string]: string[] },
  );

function getValue(obj: unknown, path: string) {
  return path.split(".").reduce(function (prev, curr) {
    return prev ? (prev as any)[curr] : null;
  }, obj);
}

export const magicResult = (
  builds: Build[],
  showInProgress: boolean = false,
  ...filter: string[]
) => {
  const possibleResults = ["failed"];
  if (showInProgress === true) {
    possibleResults.push("running");
  }

  const objFilter = objectivedFilter(...filter);
  objFilter.result = [...(objFilter.result ?? []), ...possibleResults];

  return builds.filter((item: Build) => {
    const compare = (key: string) => (val: string) =>
      getValue(item, key) === val;
    if (
      Object.entries(objFilter).every(([key, values]) =>
        values.some(compare(key)),
      )
    )
      return true;
    return false;
  });
};
