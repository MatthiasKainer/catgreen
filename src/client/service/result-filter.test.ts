import { expect, test } from "vitest";
import { magicResult as magicFilter, objectivedFilter } from "./result-filter";
import response from "../test-utils/response.json";
import { Build } from "../../domain/build";

test("it will give only failing results if nothing else set", () => {
  const results = magicFilter(response.data as Build[]);
  expect(results).toMatchSnapshot();
  expect(results.every(({ result }) => result === "failed")).toBe(true);
});

test("it will give failing and running results if specified", () => {
  const results = magicFilter(response.data as Build[], true);
  expect(results).toMatchSnapshot();
  expect(results.length).toBe(20);
  expect(
    results.every(({ result }) => result === "failed" || result === "running"),
  ).toBe(true);
});

test("it will create an objectived filter correctly", () => {
  expect(objectivedFilter()).toStrictEqual({});
  expect(objectivedFilter(":::")).toStrictEqual({});
  expect(objectivedFilter("mauahaha")).toStrictEqual({});
  expect(objectivedFilter("meta.blame:doc-orc")).toEqual({
    "meta.blame": ["doc-orc"],
  });
  expect(objectivedFilter(" meta.blame:doc-orc ")).toEqual({
    "meta.blame": ["doc-orc"],
  });
  expect(objectivedFilter("meta.blame : doc-orc ")).toEqual({
    "meta.blame": ["doc-orc"],
  });
  expect(
    objectivedFilter("meta.blame:doc-orc", "meta.blame:grumpy-orc"),
  ).toEqual({ "meta.blame": ["doc-orc", "grumpy-orc"] });
  expect(
    objectivedFilter(
      "meta.blame:doc-orc",
      "meta.blame:grumpy-orc",
      "name:dopey-dragon",
    ),
  ).toEqual({
    "meta.blame": ["doc-orc", "grumpy-orc"],
    name: ["dopey-dragon"],
  });
});

test("it will filter correctly", () => {
  let result = magicFilter(
    response.data as Build[],
    true,
    "name:sleepy-codebase-ng",
  );
  expect(result.length).toBe(1);
  expect(result.every(({ name }) => name === "sleepy-codebase-ng")).toBe(true);

  result = magicFilter(response.data as Build[], true, "meta.blame:doc-orc");
  expect(result.length).toBe(1);
  expect(result.every(({ meta }) => meta.blame === "doc-orc")).toBe(true);

  result = magicFilter(
    response.data as Build[],
    true,
    "meta.blame:doc-orc",
    "result:success",
  );
  expect(result.length).toBe(12);
  expect(result.every(({ meta }) => meta.blame === "doc-orc")).toBe(true);
});
