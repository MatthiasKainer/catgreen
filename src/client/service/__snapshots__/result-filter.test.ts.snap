// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it will give failing and running results if specified 1`] = `
[
  {
    "id": "102746",
    "meta": {
      "blame": "doc-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-10T13:06:05.8690139Z",
      "triggerReason": "We jump book.",
      "when": "20 days ago",
    },
    "name": "sleepy-codebase-ng",
    "result": "failed",
  },
  {
    "id": "102747",
    "meta": {
      "blame": "sneezy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-10T13:06:08.4783945Z",
      "triggerReason": "They jump apple.",
      "when": "20 days ago",
    },
    "name": "fantastic-workspace-for-beginners",
    "result": "failed",
  },
  {
    "id": "102455",
    "meta": {
      "blame": "dopey-hobbit",
      "link": "https://example.com",
      "timestamp": "2023-11-09T14:25:37.3902661Z",
      "triggerReason": "He run book.",
      "when": "21 days ago",
    },
    "name": "awesome-ninja-final-final",
    "result": "failed",
  },
  {
    "id": "106730",
    "meta": {
      "blame": "happy-hobbit",
      "link": "https://example.com",
      "triggerReason": "You sleep car.",
      "when": "now",
    },
    "name": "awesome-ninja",
    "result": "running",
  },
  {
    "id": "106727",
    "meta": {
      "blame": "sleepy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-30T14:28:56.8787263Z",
      "triggerReason": "We jump car.",
      "when": "6 minutes ago",
    },
    "name": "creative-robot-for-beginners",
    "result": "failed",
  },
  {
    "id": "100919",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-10-31T17:29:45.1842892Z",
      "triggerReason": "He eat car.",
      "when": "29 days ago",
    },
    "name": "awesome-dragon-final",
    "result": "failed",
  },
  {
    "id": "106719",
    "meta": {
      "blame": "sneezy-orc",
      "link": "https://example.com",
      "triggerReason": "She run book.",
      "when": "now",
    },
    "name": "techy-repository-beta",
    "result": "running",
  },
  {
    "id": "106721",
    "meta": {
      "blame": "dopey-orc",
      "link": "https://example.com",
      "triggerReason": "We sleep apple.",
      "when": "now",
    },
    "name": "techy-robot-for-beginners",
    "result": "running",
  },
  {
    "id": "106485",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-29T13:31:56.8268595Z",
      "triggerReason": "We run apple.",
      "when": "1 day ago",
    },
    "name": "bashful-unicorn-ng",
    "result": "failed",
  },
  {
    "id": "101996",
    "meta": {
      "blame": "sleepy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-07T13:13:37.5439109Z",
      "triggerReason": "We eat book.",
      "when": "23 days ago",
    },
    "name": "dopey-octocat-final",
    "result": "failed",
  },
  {
    "id": "106469",
    "meta": {
      "blame": "sneezy-hobbit",
      "link": "https://example.com",
      "timestamp": "2023-11-29T13:44:54.8909092Z",
      "triggerReason": "He run book.",
      "when": "1 day ago",
    },
    "name": "techy-ninja-for-beginners",
    "result": "failed",
  },
  {
    "id": "104910",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-21T13:19:31.0352885Z",
      "triggerReason": "I eat apple.",
      "when": "9 days ago",
    },
    "name": "dopey-repository-beta",
    "result": "failed",
  },
  {
    "id": "105166",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-22T10:07:39.0068457Z",
      "triggerReason": "She run apple.",
      "when": "8 days ago",
    },
    "name": "dopey-dragon-v2",
    "result": "failed",
  },
  {
    "id": "106316",
    "meta": {
      "blame": "sleepy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-29T08:41:55.7954449Z",
      "triggerReason": "They sleep car.",
      "when": "1 day ago",
    },
    "name": "creative-ninja-ng",
    "result": "failed",
  },
  {
    "id": "102173",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-08T09:50:00.6584206Z",
      "triggerReason": "You run apple.",
      "when": "22 days ago",
    },
    "name": "sneezy-octopus-final-final",
    "result": "failed",
  },
  {
    "id": "106731",
    "meta": {
      "blame": "happy-dwarf",
      "link": "https://example.com",
      "triggerReason": "We eat apple.",
      "when": "now",
    },
    "name": "sleepy-unicorn-plus-plus",
    "result": "running",
  },
  {
    "id": "104462",
    "meta": {
      "blame": "sneezy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:33:56.6075473Z",
      "triggerReason": "She eat apple.",
      "when": "10 days ago",
    },
    "name": "awesome-project-for-beginners",
    "result": "failed",
  },
  {
    "id": "104464",
    "meta": {
      "blame": "sleepy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:34:35.0285379Z",
      "triggerReason": "He eat apple.",
      "when": "10 days ago",
    },
    "name": "creative-dragon",
    "result": "failed",
  },
  {
    "id": "104465",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:35:29.505735Z",
      "triggerReason": "We sleep apple.",
      "when": "10 days ago",
    },
    "name": "grumpy-repository-ng",
    "result": "failed",
  },
  {
    "id": "104466",
    "meta": {
      "blame": "dopey-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:39:17.6071824Z",
      "triggerReason": "I sleep car.",
      "when": "10 days ago",
    },
    "name": "bashful-unicorn-v2",
    "result": "failed",
  },
]
`;

exports[`it will give only failing results if nothing else set 1`] = `
[
  {
    "id": "102746",
    "meta": {
      "blame": "doc-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-10T13:06:05.8690139Z",
      "triggerReason": "We jump book.",
      "when": "20 days ago",
    },
    "name": "sleepy-codebase-ng",
    "result": "failed",
  },
  {
    "id": "102747",
    "meta": {
      "blame": "sneezy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-10T13:06:08.4783945Z",
      "triggerReason": "They jump apple.",
      "when": "20 days ago",
    },
    "name": "fantastic-workspace-for-beginners",
    "result": "failed",
  },
  {
    "id": "102455",
    "meta": {
      "blame": "dopey-hobbit",
      "link": "https://example.com",
      "timestamp": "2023-11-09T14:25:37.3902661Z",
      "triggerReason": "He run book.",
      "when": "21 days ago",
    },
    "name": "awesome-ninja-final-final",
    "result": "failed",
  },
  {
    "id": "106727",
    "meta": {
      "blame": "sleepy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-30T14:28:56.8787263Z",
      "triggerReason": "We jump car.",
      "when": "6 minutes ago",
    },
    "name": "creative-robot-for-beginners",
    "result": "failed",
  },
  {
    "id": "100919",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-10-31T17:29:45.1842892Z",
      "triggerReason": "He eat car.",
      "when": "29 days ago",
    },
    "name": "awesome-dragon-final",
    "result": "failed",
  },
  {
    "id": "106485",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-29T13:31:56.8268595Z",
      "triggerReason": "We run apple.",
      "when": "1 day ago",
    },
    "name": "bashful-unicorn-ng",
    "result": "failed",
  },
  {
    "id": "101996",
    "meta": {
      "blame": "sleepy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-07T13:13:37.5439109Z",
      "triggerReason": "We eat book.",
      "when": "23 days ago",
    },
    "name": "dopey-octocat-final",
    "result": "failed",
  },
  {
    "id": "106469",
    "meta": {
      "blame": "sneezy-hobbit",
      "link": "https://example.com",
      "timestamp": "2023-11-29T13:44:54.8909092Z",
      "triggerReason": "He run book.",
      "when": "1 day ago",
    },
    "name": "techy-ninja-for-beginners",
    "result": "failed",
  },
  {
    "id": "104910",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-21T13:19:31.0352885Z",
      "triggerReason": "I eat apple.",
      "when": "9 days ago",
    },
    "name": "dopey-repository-beta",
    "result": "failed",
  },
  {
    "id": "105166",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-22T10:07:39.0068457Z",
      "triggerReason": "She run apple.",
      "when": "8 days ago",
    },
    "name": "dopey-dragon-v2",
    "result": "failed",
  },
  {
    "id": "106316",
    "meta": {
      "blame": "sleepy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-29T08:41:55.7954449Z",
      "triggerReason": "They sleep car.",
      "when": "1 day ago",
    },
    "name": "creative-ninja-ng",
    "result": "failed",
  },
  {
    "id": "102173",
    "meta": {
      "blame": "grumpy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-08T09:50:00.6584206Z",
      "triggerReason": "You run apple.",
      "when": "22 days ago",
    },
    "name": "sneezy-octopus-final-final",
    "result": "failed",
  },
  {
    "id": "104462",
    "meta": {
      "blame": "sneezy-orc",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:33:56.6075473Z",
      "triggerReason": "She eat apple.",
      "when": "10 days ago",
    },
    "name": "awesome-project-for-beginners",
    "result": "failed",
  },
  {
    "id": "104464",
    "meta": {
      "blame": "sleepy-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:34:35.0285379Z",
      "triggerReason": "He eat apple.",
      "when": "10 days ago",
    },
    "name": "creative-dragon",
    "result": "failed",
  },
  {
    "id": "104465",
    "meta": {
      "blame": "dopey-dwarf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:35:29.505735Z",
      "triggerReason": "We sleep apple.",
      "when": "10 days ago",
    },
    "name": "grumpy-repository-ng",
    "result": "failed",
  },
  {
    "id": "104466",
    "meta": {
      "blame": "dopey-elf",
      "link": "https://example.com",
      "timestamp": "2023-11-20T13:39:17.6071824Z",
      "triggerReason": "I sleep car.",
      "when": "10 days ago",
    },
    "name": "bashful-unicorn-v2",
    "result": "failed",
  },
]
`;
